post-repo-inmem/src/main/java/klon/post/repo/inmem/PostRepositoryInmemImpl.java

```diff
@@ -0,0 +1,52 @@
+package klon.post.repo.inmem;
+
+import klon.post.repo.api.Post;
+import klon.post.repo.api.PostRepository;
+import lombok.Builder;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+@Builder
+public class PostRepositoryInmemImpl implements PostRepository {
+
+    private final ConcurrentHashMap<String, List<InmemPost>> posts = new ConcurrentHashMap<>();
+
+    private final PostMapper mapper;
+
+    @Override
+    public void addPost(Post post) {
+        InmemPost inmemPost = mapper.toInmemPost(post);
+        posts.merge(post.getUserId(), Collections.synchronizedList(new ArrayList<>(Collections.singletonList(inmemPost))),
+                (inmemPosts, inmemPosts2) -> {
+                    inmemPosts.addAll(inmemPosts2);
+                    return Collections.synchronizedList(inmemPosts);
```

Comment:
Czy tu się nie pojawia owijanie synchronizowanej listy kolejną synchronizowaną listą itd dla każdego nowego postu? Wydaje mi się, że po linii 23 wystarczy zwrócić inmemPosts.
Dodatkowo do rozważenie czy ArrayLista jest dobrym pomysłem. Jeśli ma urosnąć, tzn powiększyć swój rozmiar, to na nowo jest alokowana pamięć i zawartość jest przepisywana w nowe miejsce - tablica na której jest zbudowana ArrayLista musi być ciągłym obszarem pamięci. Wchodzi tu kwestia wydajności. Być może LinkedList (też ma swoje minusy)... nie wiem. Jak zawsze - ważne by wiedzieć z czym się ma do czynienia i podjąć świadomą decyzję.

Reply:
No i zastanawiam się, czy tu w ogóle jest potrzebna synchronizacja tej listy, bo synchronizację dla userId, wydaje mi się, zapewnia nam metoda merge ConcurentMapy. Dodatkowo nigdzie indziej nie widzę w kodzie synchronizacji na tej liście. Do sprawdzenia co się stanie gdy jeden wątek będzie robił stream na liście (metoda getXXX), a drugi dodawał do tej listy.

Reply:
> Czy tu się nie pojawia owijanie synchronizowanej listy kolejną synchronizowaną listą itd dla każdego nowego postu? Wydaje mi się, że po linii 23 wystarczy zwrócić inmemPosts.

Hm chyba faktycznie masz rację.

Reply:
> Być może LinkedList (też ma swoje minusy)... nie wiem

Pewnie trzeba sobie zadać standardowe pytanie czy częściej będziemy dodawać te posty czy częściej będziemy jednak szukać jakiegoś konkretnego w kolekcji. Nie wiem czy aktualnie na tym poziomie zaawansowania warto jest się na tym skupiać

Reply:
> Do sprawdzenia co się stanie gdy jeden wątek będzie **czytał ze streama** zwróconego przez metodę getXXX, a drugi w tym czasie doda do tej listy

Masz rację, bez kopiowania kolekcji bywa różnie - sprawdziłem w unit teście. Pójdę za radą i póki co zrobię streama z kopii i to wtedy działa - dodam unit test to potwierdzający.

Reply:
> Do sprawdzenia co się stanie gdy jeden wątek będzie **czytał ze streama** zwróconego przez metodę getXXX, a drugi w tym czasie doda do tej listy

Sprawdziłem unit testem faktycznie się wywalało. Pójdę za sugestią i zrobię streama na kopii. Potwierdzone unit testem, że działa poprawnie.